src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,462 @@
+                                            update_hierarchy_params());
+    }
+  }
+
+  if (!split) {
+    log_ratio_prior_prob = -log_ratio_prior_prob;
+  }
+
+  log_ratio_likelihoods =
+      pred_lpdf_divided_clust.sum() - pred_lpdf_united_clust;
+  if (!split) {
+    log_ratio_likelihoods = -log_ratio_likelihoods;
+  }
+}
+
+void SplitAndMergeAlgorithm::split_or_merge(
+    const unsigned int first_random_idx,
+    const unsigned int second_random_idx) {
+  if (allocations[first_random_idx] == allocations[second_random_idx]) {
+    split(first_random_idx, second_random_idx);
+  } else {
+    merge(first_random_idx, second_random_idx);
+  }
+}
+
+void SplitAndMergeAlgorithm::split(const unsigned int first_random_idx,
+                                   const unsigned int second_random_idx) {
+  double log_ratio_transition_prob = -restricted_gibbs_sampling(true);
+
+  double log_ratio_likelihoods = 0;
+  double log_ratio_prior_prob = 0;
+  compute_log_ratio_like_and_prior(first_random_idx, second_random_idx, true,
+                                   log_ratio_prior_prob,
+                                   log_ratio_likelihoods);
+
+  const double AcRa =
+      std::min(1.0, std::exp(log_ratio_transition_prob + log_ratio_prior_prob +
+                             log_ratio_likelihoods));
+  if (accepted_proposal(AcRa)) {
+    proposal_update_allocations(first_random_idx, second_random_idx, true);
+  }
+}
+
+void SplitAndMergeAlgorithm::merge(const unsigned int first_random_idx,
+                                   const unsigned int second_random_idx) {
+  double log_ratio_transition_prob = 0;
+  // Fake Gibbs Sampling in order to compute the transition probability
+  for (unsigned int k = 0; k < restricted_gs_data_idx.size(); k++) {
+    restricted_gs_unique_values[allocations_restricted_gs[k]]->remove_datum(
+        restricted_gs_data_idx[k], data.row(restricted_gs_data_idx[k]),
+        update_hierarchy_params());
+
+    Eigen::VectorXd logprobas(2);
+
+    if (restricted_gs_unique_values[0]->get_card() >= 1 and
+        restricted_gs_unique_values[1]->get_card() >= 1) {
+      logprobas(0) = restricted_gs_unique_values[0]->conditional_pred_lpdf(
+          data.row(restricted_gs_data_idx[k]));
+      logprobas(1) = restricted_gs_unique_values[1]->conditional_pred_lpdf(
+          data.row(restricted_gs_data_idx[k]));
+    } else {
+      if (restricted_gs_unique_values[1]->get_card() == 0) {
+        logprobas(1) = restricted_gs_unique_values[1]->prior_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+        logprobas(0) = restricted_gs_unique_values[0]->conditional_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+      } else {
+        logprobas(0) = restricted_gs_unique_values[0]->prior_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+        logprobas(1) = restricted_gs_unique_values[1]->conditional_pred_lpdf(
+            data.row(restricted_gs_data_idx[k]));
+      }
+    }
```

Comment:
By looking at the code we found out that, in this case, the lines 285-296 (the `else` branch) will never be executed, because `restricted_gs_unique_values[0]` will always contain the point `first_random_idx` and `restricted_gs_unique_values[1]` will always contain the point `second_random_idx`, because in the article on Split and Merge it is explicitly said these two points must not be considered when we do the restricted Gibbs Sampling, so they will never be moved from their initial position.
Therefore, we will proceed in removing these lines, so we can't put the `//FIXME` there.

Is it ok if we put it in line [177](https://github.com/vnardi/bayesmix/blob/fec8db54494b747eed5695ea43d04df82fa484b3/src/algorithms/split_and_merge_algorithm.cc#L177), where there's a if/else with a similar meaning?

Reply:
sure, thanks